### ISSUE TITLE
Need for non-UI end-to-end testing

### ISSUE THREAD

**Issue by simark** [eclipse-theia/theia#2919]
I recently wanted to make an end-to-end test involving Theia and clangd, to verify that some actions in the frontend (changing build configuration) worked correctly with clangd and that the system as a whole gave good results.

The only way right now is to do a UI test, where you simulate clicks in a Theia application.  However, there is nothing in this test that inherently needs to be UI-driven.  We only need some non-UI components of the frontend, a backend, and clangd.

I think it would be nice to be able to run tests where we instantiate some non-UI components of the frontend (including the bits necessary to connect to a backend) and drive the test from a standard mocha test file.

**Comment by svenefftinge** [comment]
Can't you just try to make it work and then we see how we can extract general parts if it's worth it?

**Comment by simark** [comment]
I tried a little bit and hit some road bumps, so I wanted to see if any of you saw any reason why this would be a bad idea, before going further.

**Comment by amiramw** [comment]
I did some integration non UI testings for the fontend modules of theia extensions (no backend modules). 
In order to allow tests to access them I had to expose inversify internal dictionary by injecting the following code next to the initialization of the inversify container:
```
window.TheiaApiTest = {
    getServiceBindingByClass: (s) => {
        let filteredEntries = Array.from(container._bindingDictionary.getMap().entries())
            .filter(arr => arr[0].toString().startsWith("function " + s));
        let bindingPair = filteredEntries && filteredEntries[0];
        return bindingPair && bindingPair[1] && bindingPair[1][0] && bindingPair[1][0].cache;
    }
};
```

For my initial use case this method was enough but I guess that in time there will be more of them.

I used jest-puppeteer to execute code in the browser context in my tests. So the test code looks somewhat like this:

```
let value = await page.waitForFunction(() => {
      return TheiaApiTest.getServiceBindingByClass("MyContribution")
        .doSomething();
    });
expect(value).toEqual("hello");
```

If this mechanism is desired inside theia I can create a PR.

**Comment by marechal-p** [comment]
@amiramw is it similar to https://github.com/eclipse-theia/theia/pull/5654?

**Comment by amiramw** [comment]
As far as I understand yes. But I didn't get into it yet.

**Comment by vince-fugnitto** [comment]
Closed in #6852 :)
